Data/ByteString/Char8.hs

```diff
@@ -1,7 +1,12 @@
+{-# LANGUAGE AllowAmbiguousTypes #-}
```

Comment:
(I'll hate myself for this suggestion, when I'll need to backport something to `bytestring-0.11`, but...)

Let's move `read{Word,Int}`-related functionality into a separate internal module, so that the main one remains unpolluted by new language extensions.

Reply:
OK, works for me.  I don't think it needs to be visible to users, so it should be no problem...

Reply:
> Let's move `read{Word,Int}`-related functionality into a separate internal module, so that the main one remains unpolluted by new language extensions.

I don't mind how this has played out for this PR, but what's the motivation for not using `TypeApplications` in `D.B.Char8`?